internal/controllers/machineset/machineset_controller.go

```diff
@@ -1578,3 +1573,23 @@ func (r *Reconciler) reconcileExternalTemplateReference(ctx context.Context, clu
 
 	return false, patchHelper.Patch(ctx, obj)
 }
+
+// Returns the machines to be remediated in the following order
+// Machines with RemediateMachineAnnotation annotation if any,
+// Machines failing to come up first because
+// there is a chance that they are not hosting any workloads (minimize disruption).
+func getMachinesToRemediateInOrder(remediateMachines []*clusterv1.Machine) []*clusterv1.Machine {
+	// From machines to remediate select the machines with RemediateMachineAnnotation annotation.
+	annotatedMachines := collections.FromMachines(remediateMachines...).Filter(collections.HasAnnotationKey(clusterv1.RemediateMachineAnnotation))
+
+	// Filter out the machines which are unique (machines which are not in annotated machines list)
+	uniqueMachines := collections.FromMachines(remediateMachines...).Difference(annotatedMachines).UnsortedList()
+
+	// Sort the machines from newest to oldest.
+	sort.SliceStable(uniqueMachines, func(i, j int) bool {
+		return uniqueMachines[i].CreationTimestamp.After(uniqueMachines[j].CreationTimestamp.Time)
+	})
+
+	// combine the annotated machines along with machines to remediate.
+	return append(annotatedMachines.UnsortedList(), uniqueMachines...)
+}
```

Comment:
What about this alternative implementation
```suggestion
// Returns the machines to be remediated in the following order
//   - Machines with RemediateMachineAnnotation annotation if any,
//   - Machines failing to come up first because
//     there is a chance that they are not hosting any workloads (minimize disruption).
func sortMachinesToRemediate(machines []*clusterv1.Machine) {
	sort.SliceStable(machines, func(i, j int) bool {
		_, iHasRemediateAnnotation := machines[i].Annotations[clusterv1.RemediateMachineAnnotation]
		_, jHasRemediateAnnotation := machines[j].Annotations[clusterv1.RemediateMachineAnnotation]

		if iHasRemediateAnnotation && !jHasRemediateAnnotation {
			return true
		}
		if !iHasRemediateAnnotation && jHasRemediateAnnotation {
			return false
		}

		return machines[i].CreationTimestamp.After(machines[j].CreationTimestamp.Time)
	})
}
```

(trying to have something easier to read/that sticks to the fact that this is a sorting func)

Reply:
Yeah makes sense, I think I took the complicated way, I just updated as suggested but still kept in a seperate function to easily test, Let me know if I need to make it inplace and enhance the existing test itself.